### ISSUE TITLE
Add a TryLoadingLibraryFiles method

### ISSUE THREAD

**Issue by Symbai** [Ruslan-B/FFmpeg.AutoGen#175]
* **I'm submitting a ... **
  - [ ] bug report
  - [X] feature request
  - [ ] support request or question => Please do not submit support request or questions here, see note at the top of this template.

* **What is the current behavior?***
For windows and Linux and maybe MacOS there can be different folders where to lookout for the library files (Current folder, PATH system environment variable, fixed path under Linux, sub folder, etc.pp). Currently we can either enumerate the files ourselves on all of these places before setting the RootPath property (but `ffmpeg.autogen` does not provide the full filenames directly to us) or simply set it and hope no DllNotFound exception raises. See also https://github.com/Ruslan-B/FFmpeg.AutoGen/blob/master/FFmpeg.AutoGen.Example/FFmpegBinariesHelper.cs there is no property validation that tells us if the library files really exist or not.

* **What is the expected behavior?**
I suggest adding a simple helper method that returns a boolean if all library files exist. Preferably if all library files exist they also get loaded:

```C#
public bool TryLoadingLibraryFiles(string? path = null) {
	path ??= RootPath;
	bool allFound = false;
	foreach (var item in LibraryDependenciesMap) {
		//if File.Exists(...)
	}
	return allFound;
}
```

* **What is the motivation / use case for changing the behavior?**
We can conveniently write something like this to look at multiple paths:
```C#
string rootPath = ffmpeg.RootPath;
if (!ffmpeg.TryLoadingLibraryFiles()) {
	if (!ffmpeg.TryLoadingLibraryFiles(Path.Combine(rootPath, "bin"))) {
		ffmpeg.RootPath = Path.Combine(rootPath, "ffmpeg");
		//....
	}
}
```

* **Please tell us about your environment:**

- version: 4.3.0.3

**Comment by Ruslan-B** [comment]
I have struggle to understand rationale behind this as you have access to ffmpeg.RootPath. Can you elaborate?

**Comment by Symbai** [comment]
As written in first post:

> For windows and Linux and maybe MacOS there can be different folders where to lookout for the library files (Current folder, PATH system environment variable, fixed path under Linux, sub folder, etc.pp)

`ffmpeg.RootPath` might be invalid. It says it cannot find ffmpeg because it doesn't look at all possible paths. Adding a helper method makes it easier for us to look at all possible paths. If there is something specific you don't understand please point to it.

**Comment by Ruslan-B** [comment]
We are speaking same language or not? To make clear - it is not responsibility of the library - as code you are referring to is example - to find correct location of the binaries. I think you have all the tools to make as flexible as you like too.

**Comment by Speykious** [comment]
> To make clear - it is not responsibility of the library [...] to find correct location of the binaries.

I'm pretty sure that is not what they meant. The problem is that currently, there is no explicit way to *try* loading libraries given the current root path. The only available option would be to call an `ffmpeg` function that has to load the library and see if it throws a `DllNotFoundException`.

If we simply want to try a bunch of paths to get one that works, this technique is an absolute hack. It would be much simpler and efficient if a `TryLoadingLibraryFiles` function existed, tried to find all the libraries and returned a boolean, or maybe fr more advanced error management, a dictionary of required libraries each with a boolean indicating if they have been found or not.

And that is not asking `FFmpeg.Autogen` to be responsible of finding the correct location of the native libraries, but merely to *provide an easy way for users of the libraries do so*.

Thus I believe that it is definitely in this library's interest to provide such a function.

**Comment by Ruslan-B** [comment]
Okay, finally rationale, got it I'll see what can we can improve here.

**Comment by Ruslan-B** [comment]
I cannot accept #181

**Comment by Maltragor** [comment]
Maybe I didn't quite understand the problem, but what would speak against the following approach (as part of the application)?
If you don't know if the libraries are available and where are they located, why should you look for them yourself when the OS does it for you. (And on the other hand it would of course be possible to specify a root path to search specifically)

```C#
public bool TryLoadingLibraryFiles() {
	try {
		ffmpeg.RootPath = "";
		foreach (KeyValuePair<string, int> item in ffmpeg.LibraryVersionMap)
			ffmpeg.GetOrLoadLibrary(item.Key);
		return true;
	}
	catch { }
	return false;
}
```
ffmpeg.GetOrLoadLibrary(libraryName)-> ... -> LibraryLoader.LoadNativeLibrary(RootPath, libraryName,..) -> 
a) LinuxNativeMethods.dlopen(Path.Combine(path, nativeLibraryName),...) -> dlopen(filename)
b) WindowsNativeMethods.LoadLibrary(filename)
=> If filename does not contain any path, the OS specific dynamic linker search the files based on multiple strategies.
At least under Linux the approach had worked for me. (Other platforms not tested)

**Comment by Ruslan-B** [comment]
@Maltragor it wont'n work in pure manner as libraries might be loaded from different locations and never being unloaded again.
I am strictly agains forcibly unloading them as it is asking for instability just from the beginning - normal dotnet interop doesn't allow you to do so.
Similar function which can check complaint file set is a way better way from my perspective.

**Comment by Speykious** [comment]
@Maltragor See pull requests above. The first one I made (#181) was doing that, with the possibility of trying out different paths and being able to specify which libraries from FFmpeg are needed. It got rejected for the reason @Ruslan-B mentioned above.

The second one I made (#183) was essentially the same thing, except it only checked if the files were there without loading them, which is what @Ruslan-B calls the better way. But it also got rejected unfortunately.
